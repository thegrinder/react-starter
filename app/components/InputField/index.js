import React from 'react';
import PropTypes from 'prop-types';
import { Input, Text } from 'basic-styled-uikit';

const propTypes = {
  onChange: PropTypes.func.isRequired,
  onBlur: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  touched: PropTypes.bool.isRequired,
  id: PropTypes.string.isRequired,
  label: PropTypes.string,
  error: PropTypes.string,
};

const defaultProps = {
  error: '',
  label: '',
};


const InputField = ({
  label,
  onChange,
  onBlur,
  value,
  error,
  touched,
  id,
  ...rest
}) => (
  <div className="mb-2">
    {label && <Text as="label" htmlFor={id}>{label}</Text>}
    <Input
      id={id}
      onChange={onChange}
      onBlur={onBlur}
      invalid={!!error && touched}
      value={value}
      {...rest}
    />
    <div className="h-4">
      {touched && !!error && (
        <Text sizing="s" color="danger" className="truncated">
          {error}
        </Text>
      )}
    </div>
  </div>
);

InputField.propTypes = propTypes;
InputField.defaultProps = defaultProps;

export default InputField;
