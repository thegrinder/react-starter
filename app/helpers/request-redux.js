export const initialRequestState = {
  error: null,
  initialLoad: true,
  loading: false,
};

export const createRequest = (namespace) => {
  const actionTypes = {
    FAILED: `${namespace}/FAILED`,
    FULFILLED: `${namespace}/FULFILLED`,
    LOADING: `${namespace}/LOADING`,
    RESET: `${namespace}/RESET`,
  };

  const actionCreators = {
    failed: error => ({ error, type: actionTypes.FAILED }),
    fulfilled: () => ({ type: actionTypes.FULFILLED }),
    loading: () => ({ type: actionTypes.LOADING }),
    reset: () => ({ type: actionTypes.RESET }),
  };

  const reducer = (state = initialRequestState, action) => {
    switch (action.type) {
      case actionTypes.LOADING:
        return {
          ...state,
          error: null,
          loading: true,
        };
      case actionTypes.FAILED:
        return {
          ...state,
          error: action.error,
        };
      case actionTypes.FULFILLED:
        return {
          ...state,
          initialLoad: false,
          loading: false,
        };
      case actionTypes.RESET:
        return initialRequestState;
      default:
        return state;
    }
  };

  return {
    actionCreators,
    actionTypes,
    reducer,
  };
};
