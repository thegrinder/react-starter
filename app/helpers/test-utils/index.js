import React from 'react';
import PropTypes from 'prop-types';
import { IntlProvider } from 'react-intl';
import { ThemeProvider } from 'styled-components';
import { render } from 'react-testing-library';

import theme from '../../theme/components';

const propTypes = {
  children: PropTypes.node.isRequired,
};

const Providers = ({ children }) => (
  <IntlProvider locale="en">
    <ThemeProvider theme={theme}>
      {children}
    </ThemeProvider>
  </IntlProvider>
);

Providers.propTypes = propTypes;

const customRender = (ui, options) => render(ui, { wrapper: Providers, ...options });

export * from 'react-testing-library';
export { customRender as render };
