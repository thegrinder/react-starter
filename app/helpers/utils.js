export const rem = pxValue => `${pxValue / 16}rem`;

export const isEmptyObj = obj => Object.keys(obj).length === 0;

export const isEmptyArr = arr => arr.length === 0;

export const union = (arr1, arr2) => arr1.concat(arr2.filter(i => !arr1.includes(i)));

export const omitBy = (obj, properties) => (
  Object.keys(obj)
    .filter(property => !properties.includes(property))
    .reduce(
      (prev, next) => ({
        ...prev,
        [next]: obj[next],
      }),
      {},
    )
);
