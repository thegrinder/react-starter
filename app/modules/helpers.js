export const initialRequestState = {
  error: {},
  initialLoad: true,
  loading: false,
  succeeded: false,
};

export const createRequest = namespace => {
  const actionTypes = {
    FAILED: `${namespace}/FAILED`,
    FULFILLED: `${namespace}/FULFILLED`,
    LOADING: `${namespace}/LOADING`,
    RESET: `${namespace}/RESET`,
    UPDATE: `${namespace}/UPDATE`,
    SUCCEEDED: `${namespace}/SUCCEEDED`,
    TRIGGER: `${namespace}/TRIGGER`,
  };

  const actionCreators = {
    failed: error => ({ error, type: actionTypes.FAILED }),
    succeeded: () => ({ type: actionTypes.SUCCEEDED }),
    fulfilled: () => ({ type: actionTypes.FULFILLED }),
    loading: () => ({ type: actionTypes.LOADING }),
    reset: () => ({ type: actionTypes.RESET }),
  };

  const reducer = (state = initialRequestState, action) => {
    switch (action.type) {
      case actionTypes.LOADING:
        return {
          ...state,
          error: {},
          loading: true,
          succeeded: false,
        };
      case actionTypes.SUCCEEDED:
        return {
          ...state,
          succeeded: true,
        };
      case actionTypes.FAILED:
        return {
          ...state,
          error: action.error,
        };
      case actionTypes.FULFILLED:
        return {
          ...state,
          initialLoad: false,
          loading: false,
        };
      case actionTypes.RESET:
        return {
          ...initialRequestState,
          initialLoad: state.initialLoad,
        };
      default:
        return state;
    }
  };

  return {
    actionCreators,
    actionTypes,
    reducer,
  };
};
