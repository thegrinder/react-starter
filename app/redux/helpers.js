export const initialRequestState = {
  error: {},
  initialLoad: true,
  loading: false,
};

export const createRequest = (namespace) => {
  const actionTypes = {
    FAILED: `${namespace}/FAILED`,
    FULFILLED: `${namespace}/FULFILLED`,
    LOADING: `${namespace}/LOADING`,
    RESET: `${namespace}/RESET`,
    SUCCEEDED: `${namespace}/SUCCEEDED`,
    TRIGGER: `${namespace}/TRIGGER`,
  };

  const actionCreators = {
    failed: error => ({ error, type: actionTypes.FAILED }),
    fulfilled: () => ({ type: actionTypes.FULFILLED }),
    loading: () => ({ type: actionTypes.LOADING }),
    reset: () => ({ type: actionTypes.RESET }),
  };

  const reducer = (state = initialRequestState, action) => {
    switch (action.type) {
      case actionTypes.LOADING:
        return {
          ...state,
          error: {},
          loading: true,
        };
      case actionTypes.FAILED:
        return {
          ...state,
          error: action.error,
        };
      case actionTypes.FULFILLED:
        return {
          ...state,
          initialLoad: false,
          loading: false,
        };
      case actionTypes.RESET:
        return initialRequestState;
      default:
        return state;
    }
  };

  return {
    actionCreators,
    actionTypes,
    reducer,
  };
};
