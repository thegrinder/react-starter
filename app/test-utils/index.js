/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import PropTypes from 'prop-types';
import configureStore from 'redux-mock-store';
import { IntlProvider } from 'react-intl';
import { ThemeProvider } from 'styled-components';
import * as RTL from '@testing-library/react';

import theme from '../theme';

const propTypes = {
  children: PropTypes.node.isRequired,
};

const Providers = ({ children }) => (
  <IntlProvider locale="en">
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
  </IntlProvider>
);

Providers.propTypes = propTypes;

export const render = (ui, options) =>
  RTL.render(ui, { wrapper: Providers, ...options });

export const { fireEvent, act } = RTL;

export const createStore = configureStore();
