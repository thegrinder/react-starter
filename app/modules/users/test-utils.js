import { initialRequestState } from '../helpers';
import { initialUsersState } from './reducer';

export const createUser = (user = {}) => ({
  id: 'id',
  name: 'name',
  username: 'username',
  email: 'email@email.com',
  phone: '132456776',
  ...user,
});

export const createRequestState = (requestState = {}) => ({
  ...initialRequestState,
  ...requestState,
});

export const createUsersState = ({
  requestKey = '',
  requestState = {},
  data = {},
} = {}) => ({
  ...initialUsersState,
  data: {
    ...initialUsersState.data,
    ...data,
  },
  requests: {
    ...initialUsersState.requests,
    ...(requestKey && {
      [requestKey]: {
        ...initialUsersState.requests[requestKey],
        ...requestState,
      },
    }),
  },
});

export const createState = args => ({
  users: createUsersState(args),
});
