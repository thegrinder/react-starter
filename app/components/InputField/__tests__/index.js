import React from 'react';

import { render } from '../../../helpers/test-utils';
import InputField from '..';

const requiredProps = {
  onChange: jest.fn(),
  onBlur: jest.fn(),
  value: '',
  touched: false,
  id: 'id',
};


const renderComponent = (props = {}) => render(
  <InputField {...requiredProps} {...props} />,
);

describe('<InputField />', () => {
  it('should render correctly', () => {
    const { container: { firstChild } } = renderComponent();
    expect(firstChild).toBeDefined();
    expect(firstChild).toMatchSnapshot();
  });

  it('should render error and label', () => {
    const { container: { firstChild }, getByText } = renderComponent({
      touched: true,
      error: 'error',
      label: 'label',
    });
    const errorElement = getByText('error');
    const labelElement = getByText('label');
    expect(firstChild).toContainElement(errorElement);
    expect(firstChild).toContainElement(labelElement);
  });
});
