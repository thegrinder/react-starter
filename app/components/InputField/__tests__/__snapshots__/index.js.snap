// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputField /> should render correctly 1`] = `
<div
  class="mb-2"
>
  
  <input
    class="StyledInput-ok3pz6-0 buStls"
    id="id"
    value=""
  />
  <div
    class="h-4"
  />
</div>
`;
